fix(masking): Various performance, functional, and readability fixes for JSON-based masking

jamfor999 opened the pull request:
Several fixes here:
1) I realised after further usage and some extensive testing in more complex scenarios that JSON Arrays weren't handled correctly - they were always either masked at the whole field level (on Mask By Default mode), or left unmasked (on Show By Default mode). Now, the same settings that apply now also apply to nested objects inside arrays, or if the array contains primitives, the higher level array will apply. So for example, given:

```
address.firstLine
```

this will now apply to both:

```
{
  "address": {
      "firstLine": "value"
  }
}
```
and now also to all the `firstLine` values in an array type, too (this was not working correctly previously):
```
{
  "address": [
    {
      "firstLine": "value"
    },
    {
      "firstLine": "someOtherValue"
    }
  ]
}
```

Primitives will be handled as expected as well, where a filter like:

```
address.values
```

would now result that the following object:

```
{
    "address": {
      "values": [ "one", "two", "three" ]
    }
}
```

will correctly look like:

```
{
    "address": {
      "values": [ "xxxx", "xxxx", "xxxx" ]
    }
}
```

instead of the current:

```
{
    "address": {
      "values": "xxxx"
    }
}
```
which is currently losing the format, as well as non-sensitive information such as array length.

2) Some readability improvements & fixes, also updating tests to be uniform, so the exact same test code is not repeated. Now the `JsonMaskerTest` interface defines the core tests the input and output, with the `ShowByDefault` and `MaskByDefault` tests implementing the interface (with their own backing `application.yml` which makes them fit the test criteria) and have their own specific tests written in the class only

3) Small performance improvements, including defining the topic -> filtering keys in the constructor rather than constantly re-evaluating
 
4) Remove the phrase "Please contact akhq administrator" which is likely to cause confusion and bring people here, when the issue isn't AKHQ but that a record has not fulfilled the criteria for masking

jamfor999 committed:
fix(masking): Various performance and readability fixes

jamfor999 committed:
fix: Remove 'please contact akhq administrator'

jamfor999 commented:
cc: @AlexisSouquiere since you have familiarity reviewing this code beforehand :) A few things changed as explained in the PR. Makes it more reliable and functional and should improve performance (a bit) too

jamfor999 committed:
fix: Use JsonElement instead.

jamfor999 committed:
Merge branch 'dev' into dev

jamfor999 commented:
hi @tchiotludo I was wondering if I could get a quick look on this please? It would be great :)
Thank you

jamfor999 committed:
Merge branch 'dev' into dev

AlexisSouquiere commented:
@jamfor352 I'm starting to review the PR !

AlexisSouquiere commented on src/main/java/org/akhq/utils/JsonMaskByDefaultMasker.java:
```diff
@@ -1,73 +1,123 @@
-import com.google.gson.JsonElement;
-import com.google.gson.JsonObject;
-import com.google.gson.JsonParser;
+import com.google.gson.*;
```
Can you avoid using * and import only the specific types ?

AlexisSouquiere commented on src/main/java/org/akhq/utils/JsonShowByDefaultMasker.java:
```diff
@@ -1,65 +1,122 @@
+        this.topicToKeysMap = buildTopicKeysMap(dataMasking);
+    }
+
+    private Map<String, List<String>> buildTopicKeysMap(DataMasking dataMasking) {
```
This method is duplicated in JsonMaskByDefaultMasker

AlexisSouquiere commented on src/main/java/org/akhq/utils/JsonShowByDefaultMasker.java:
```diff
@@ -1,65 +1,122 @@
-        this.jsonMaskingFilters = dataMasking.getJsonFilters();
         this.jsonMaskReplacement = dataMasking.getJsonMaskReplacement();
+        this.topicToKeysMap = buildTopicKeysMap(dataMasking);
+    }
```
Can you move all the common variables / constructor upwards in Masker ?

jamfor999 replied:
> Can you move all the common variables / constructor upwards in Masker ?

Masker is an interface, but I can make an intermediary abstract class, for sure

AlexisSouquiere commented on src/main/java/org/akhq/utils/JsonMaskByDefaultMasker.java:
```diff
@@ -1,73 +1,123 @@
+        return topicToKeysMap.getOrDefault(topic.toLowerCase(), Collections.emptyList());
+    }
+
+    private Record createNonJsonRecord(Record record) {
```
These 3 simple functions have a single reference in maskRecord. I think you can remove them and do the job in maskRecord directly

jamfor999 replied:
> These 3 simple functions have a single reference in maskRecord. I think you can remove them and do the job in maskRecord directly

while true, I wanted to make the intention behind them more readable as lots of detail in one method can be difficult to parse. But I also appreciate these bits may be overly simple and add too much to the stack, so I'll adjust it

AlexisSouquiere commented on src/main/java/org/akhq/utils/JsonShowByDefaultMasker.java:
```diff
@@ -1,65 +1,122 @@
-        return record;
+    }
+
+    private Record maskJsonRecord(Record record) {
```
Merge with maskJsonRecord as it's only used here

AlexisSouquiere commented on src/main/java/org/akhq/utils/JsonShowByDefaultMasker.java:
```diff
@@ -1,65 +1,122 @@
-            if (node.has(keys[index])) {
-                node.addProperty(keys[index], jsonMaskReplacement);
-            }
+    private String[][] preProcessPaths(List<String> maskedKeys) {
```
Merge with applyMasking

AlexisSouquiere reviewed (changes requested):
Can you also add some comments to explain that you handle nested objects aa.bb but that it also applies to arrays ? For maintainability :)

jamfor352 commented:
Hi @AlexisSouquiere , just wanted to leave a comment as I haven't got round to this as I've been taking some time to rest after a period of illness but I'll try my best to get round to these comments pretty soon. Hopefully within the next few days, assuming I'm better.

jamfor999 committed:
fix: Response to PR

jamfor999 commented:
THanks for the review @AlexisSouquiere , I've addressed your comments :)

jamfor999 committed:
Merge branch 'dev' into dev

jamfor999 committed:
fix: More documentation and explicitly testing an array of objects.

jamfor999 commented:
Realised I missed the documentation request. I've added this, and also added more testing as well @AlexisSouquiere , so hopefully it should all be up to scratch now.

jamfor999 committed:
Merge branch 'dev' into dev

jamfor999 committed:
fix: Improve wording by making it clearer.

jamfor999 commented:
Hi @tchiotludo @AlexisSouquiere is there any chance this could be reviewed soon? My team is itching to get the array issue fixed :)

AlexisSouquiere reviewed (approved):
LGTM

jamfor999 committed:
Merge branch 'dev' into dev

jamfor999 commented:
> LGTM

Thank you! 
I think this is ready to merge now @tchiotludo , I won't be updating this PR any more as I'm happy with the state of it
